src/common/snippets/src/runtime_configurator.cpp

```diff
@@ -62,7 +64,7 @@ void RuntimeConfigurator::initialization(const lowered::LinearIRCPtr& linear_ir)
     m_latest_shapes.resize(m_io_num);
     m_config->io_data_offsets.resize(m_io_num);
     m_config->tile_rank = linear_ir->get_config().m_loop_depth;
-    m_optimizer.init(linear_ir, m_io_descs, m_in_num);
+    m_optimizer.init(linear_ir);
```

Comment:
What's about to just create the ctor of `MHAParallelWAOptimizer` with the following args: `const LinearIR& lir, const std::shared_ptr<RuntimeConfigurator>& configurator` instead of the separate method `MHAParallelWAOptimizer::init`?

Then you derive `RuntimeConfigurator` from `std::shared_from_this<RuntimeConfigurator>`, can store configurator as smart pointer in the optimizer and initialize optimizer in `initialization`:
```
void RuntimeConfigurator::initialization(const lowered::LinearIRCPtr& linear_ir) {
...
m_optimizer = std::make_shared<MHAParallelWAOptimizer>(linear_ir, this->shared_from_this());
}
```

What do you think?

Reply:
I think it is a good idea, I implemented it. However, my implementation has several nuances which misalign with your suggestion:

1. I have to create default constructor for `MHAParallelWAOptimizer` in order to initialize `m_optimizer` at `RuntimeConfigurator` constructor stage.
2. `m_optimizer` is stored in `RuntimeConfigurator` as object, not pointer.
3. I don't see any reasons for storing `RuntimeConfigurator` inside `MHAParallelWAOptimizer` as a smart pointer, and inherit `RuntimeConfigurator` from `std::enable_shared_of_this`. The optimizer is anyway stored as `RuntimeConfigurator`'s member, and it is removed during `RuntimeConfigurator`'s destructor. So the optimizer can't be used after the corresponding configurator is removed